File: mine/ch-12/arr-max.js
const arrMax = arr => {
	if (arr.length === 2) {
		return arr[0] > arr[1] ? arr[0] : arr[1];
	}

	const highest = arrMax(arr.slice(1));
	return arr[0] > highest ? arr[0] : highest;
};

console.log(arrMax([2, 6, 3, 8]));

File: mine/ch-12/arr_max.py
def arr_max(arr: list):
    print("In arr_max")
    if len(arr) == 1:
        return arr[0]

    if arr[0] > arr_max(arr[1:]):
        return arr[0]

    return arr_max(arr[1:])


print(arr_max([1, 2, 3]))


def opt_arr_max(arr: list):
    print("In opt_arr_max")
    # Optimized to not call recursion on last number so -1 on the call stack
    if len(arr) == 2:
        return arr[0] if arr[0] > arr[1] else arr[1]

    # Use var to only call recursion one time instead of two for each highest number
    # from rest of arr
    highest = opt_arr_max(arr[1:])
    return arr[0] if arr[0] > highest else highest


print(opt_arr_max([1, 2, 3, 4]))
